src/libsystemd/sd-login/sd-login.c

```diff
@@ -39,75 +39,26 @@
  *    requested metadata on object is missing → -ENODATA
  */
 
-_public_ int sd_pid_get_session(pid_t pid, char **ret_session) {
-        int r;
-
-        assert_return(pid >= 0, -EINVAL);
-        assert_return(ret_session, -EINVAL);
-
-        r = cg_pid_get_session(pid, ret_session);
-        return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;
-}
-
-_public_ int sd_pid_get_unit(pid_t pid, char **ret_unit) {
-        int r;
-
-        assert_return(pid >= 0, -EINVAL);
-        assert_return(ret_unit, -EINVAL);
-
-        r = cg_pid_get_unit(pid, ret_unit);
-        return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;
-}
-
-_public_ int sd_pid_get_user_unit(pid_t pid, char **ret_unit) {
-        int r;
-
-        assert_return(pid >= 0, -EINVAL);
-        assert_return(ret_unit, -EINVAL);
-
-        r = cg_pid_get_user_unit(pid, ret_unit);
-        return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;
-}
-
-_public_ int sd_pid_get_machine_name(pid_t pid, char **ret_name) {
-        int r;
-
-        assert_return(pid >= 0, -EINVAL);
-        assert_return(ret_name, -EINVAL);
-
-        r = cg_pid_get_machine_name(pid, ret_name);
-        return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;
-}
-
-_public_ int sd_pid_get_slice(pid_t pid, char **ret_slice) {
-        int r;
-
-        assert_return(pid >= 0, -EINVAL);
-        assert_return(ret_slice, -EINVAL);
-
-        r = cg_pid_get_slice(pid, ret_slice);
-        return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;
-}
-
-_public_ int sd_pid_get_user_slice(pid_t pid, char **ret_slice) {
-        int r;
-
-        assert_return(pid >= 0, -EINVAL);
-        assert_return(ret_slice, -EINVAL);
-
-        r = cg_pid_get_user_slice(pid, ret_slice);
-        return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;
-}
-
-_public_ int sd_pid_get_owner_uid(pid_t pid, uid_t *ret_uid) {
-        int r;
+#define DEFINE_PID_GETTER_FULL(name, type) \
+        _public_ int sd_pid_get_##name(pid_t pid, type *ret) {   \
+                int r;                                          \
+                                                                \
+                assert_return(pid >= 0, -EINVAL);               \
+                assert_return(ret, -EINVAL);                    \
+                                                                \
+                r = cg_pid_get_##name(pid, ret);                \
+                return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;    \
+        }
```

Comment:
hmm, this is used exclusively for returning strings, with a single exception. I wonder if it wouldn#t be nicer to avoid macros for this, after all macros kinda suck, and we only use them where they bring major benefits to the table. i.e. maybe:


```c
static int map_cgroup_error(int r) {
        return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;
}

static int generic_cgroup_getter(pid_t pid, int (*getter)(pid_t pid, char **ret), char **ret) {
         assert_return(pid >= 0, -EINVAL);
         assert_return(ret, -EINVAL); 

         return map_cgroup_error(getter(pid, ret));
}

static sd_pid_get_session(pid_t pid, char **ret_session) {
         return generic_cgroup_getter(pid, cg_pid_get_session, ret_session);
}

static sd_pid_get_uid(pid_t pid, uid_t **ret_uid) {
         assert_return(pid >= 0, -EINVAL);
         assert_return(ret, -EINVAL); 

         return map_cgroup_error(cg_pid_get_owner_uid(pid, ret_uid));
}
```

not quite as short as with the macros, but I think avoiding the macros here makes this much nicer.
```